# CoV-Dist 

## Overview
A python script to measure the dissimilarity distance between SARS-CoV-2 metagenome samples.

It will 
1) Calculate the Yue & Clayton dissimilarity index of two samples based on the base-pair proportions at each genomic position;
2) Sum all the dissimilarity indexes across the genome;
3) Output the sum as a distance metric.

## Prerequisites
+ [Biopython](https://biopython.org/)
+ [pysam](https://github.com/pysam-developers/pysam)(>=0.15.3)
+ [pysamstats](https://github.com/alimanfoo/pysamstats)
+ [scikit-bio](https://github.com/biocore/scikit-bio) (>=0.5.6)
+ [seaborn](https://github.com/mwaskom/seaborn)

```
# set conda environment
conda create --name cov-dist python=3.7 biopython seaborn scikit-bio pysam
conda activate cov-dist
```



## Tutorial

```
usage: covdist.py [-h] -f  -m  [-q] [-c] [-r] -o

Compute pairwise metagenome distance for SARS-CoV-2 samples

optional arguments:
  -h, --help    show this help message and exit
  -f , --file   file that lists the path to sorted bam files
  -m , --meta   file containing metadata info
  -q , --qual   Only reads with mapping quality equal to or greater than this
                value will be counted (0 by default).
  -c , --cov    Only samples with reads mapped to equal to or greater than
                this fraction of the genome will be used for PcoA analysis
                (0.5 by default).
  -r , --ref    input reference file
  -o , --out    output folder name

```

### Input
* Required: A file that specifes the path to the sorted bam files. These files should be generated by aligning SARS-CoV-2 metagenome samples to a SARS-CoV-2 reference genome.
* Optional: Reference genome used for bam alignment (default: NC_045512.2)
* Optional: Alignment quality threshold (default: 0; no filter)
* Required: A table file with two columns (tab delimited):

   Column number | Explanation                                                            |
  ---------------|------------------------------------------------------------------------|
    1            | The basename of the sample files
    2            | The category to which each sample belongs (Graph will be colored by these categories)

### Output
In the output directory, 6 files will be generated as listed below:

```
out_dir/
├── distance.txt
├── ordination_result.txt
├── PcoA_plot_3D.png
├── PcoA_plot_3D.svg
├── PcoA_plot.png
└── PcoA_plot.svg

0 directories, 6 files
```

* Calculated distance matrix: `distance.txt`
* Output from the principal coordinates analysis: `ordination_result.txt`
* PcoA plot(based on first two axes or three axes) in png and svg format:
  * `PcoA_plot_3D.png`
  * `PcoA_plot_3D.svg`
  * `PcoA_plot.png`
  * `PcoA_plot.svg`

## Notes
* If there is zero coverage for a position in the alignment, we assume the dissimilarity in that position is zero, which could lead to an underestimation of the distance between samples with low coverage.
* This method in theory could be measure dissimilarity between metagenomes and isolate genomes.
* The `distance.txt` could be used as input for other dimensionality reduction methods.
* The `ordination_result.txt` could be used as input for your preferred data visualization tool to plot the PcoA.
